include/neuralnet/layer.h

```diff
@@ -85,6 +85,112 @@ class DropoutLayer: public Layer {
 };
 
 /**
+  * DBM bottom layer
+  */
+class DBMBottomLayer: public Layer {
+ public:
+  using Layer::Setup;
+  using Layer::SetupAfterPartition;
+  using Layer::ComputeFeature;
+  using Layer::ComputeGradient;
+
+  virtual void Setup(const LayerProto& proto,
+      const vector<SLayer>& srclayers);
+  virtual bool is_bottomlayer() const {
+    return true;
+  }
+  virtual void SetupAfterPartition(const LayerProto& proto,
+      const vector<int> &shape,
+      const vector<SLayer>& srclayers);
+  virtual ConnectionType connection_type(int k) const {
+    CHECK_LT(k, srclayers_.size());
+    return kOneToAll;
+  }
+
+  virtual void ComputeFeature(Phase phase,
+     const vector<shared_ptr<Layer>>& srclayers);
+  virtual void ComputeGradient(const vector<shared_ptr<Layer>>& srclayers);
+  virtual void ComputeLoss(Metric* perf);
+  virtual Blob<float>* mutable_data(const Layer* from) {
+    if (kPhase)
```

Comment:
where is kPhase defined?
variables starts with 'k' is usually used as constants.
you may want to change mutable_data as 
... mutable_data(const Layer* from, Phase phase=kTrain)
and then check using phase. But all mutable_data function should be changed in layer.h and base_layer.h.